### DISCUSSION TITLE
RESTeasy client returns HTTP 405 while Jersey client returns HTTP 200

### DISCUSSION THREAD

**Discussion by nicolasduminil** [resteasy/resteasy#3549] [Q&A]
Hello,

I have Jakarta RESTfull endpoint deployed on Wildfly 27. Here is the code:

      @POST
      @Path("/add")
      @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
      @Produces(MediaType.TEXT_HTML)
      public Response createItem(@FormParam("key")String key, @FormParam("value")String value)
      {
        itemFacade.addToList(key,value);
        return Response.ok(RESPONSE_OK).build();
      }

I'm using the following RESTeasy client to make a POST request:

        <dependency>
          <groupId>org.jboss.resteasy</groupId>
          <artifactId>resteasy-client</artifactId>
          <version>6.2.2.Final</version>
        </dependency>

      @Test
      @Order(10)
      public void test() throws MalformedURLException
      {
        Response response = webTarget.path("add").request().accept(MediaType.TEXT_HTML).post(Entity.form(new Form()
          .param("key", "myKey").param("value", "myValue")));
        assertThat(response).isNotNull();
        assertThat(response.getStatus()).isEqualTo(HttpStatus.SC_OK);
      }

This test fails because the POST returns HTTP 405. But when I use a Jersey client instead of a RESTeasy one, using the following dependencies:

    <dependency>
      <groupId>org.glassfish.jersey.core</groupId>
      <artifactId>jersey-client</artifactId>
      <version>3.1.0</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.glassfish.jersey.inject</groupId>
      <artifactId>jersey-hk2</artifactId>
      <version>3.1.0</version>
      <scope>test</scope>
    </dependency>

then everything works as expected and the POST request returns HTTP 200.

Given that I'm deploying on Wildfly and that I'm using the wildfly-ee-with-tools BOM, I prefer to use the RESTeasy client. But for some reason I can't make it working. What am I doing wrong here ?

Many thanks in advance.

Nicolas

**Comment by jamezp** [comment]
Are you sure your `WebTarget` is setup correctly? I just tested locally and it worked fine for me.

**Reply by nicolasduminil** [reply]
The WebTarget is set as shown below, i.e. exactly the same way as it is set with the Jersey client, which works as expected.

        @ArquillianResource
        private URL url;
        private Client client;
        private WebTarget webTarget;
        ...
      
        @BeforeEach
        public void beforeEach()
        {
          client = ClientBuilder.newClient();
          webTarget = client.target(url.toString()).path("/api/kv");
        }

**Reply by jamezp** [reply]
Interesting. That looks correct to me. Have you verified the path is correct? The 405 error indicates the HTTP method isn't allowed. I'm wasn't testing this in an Arquillian test, but create the request like this is pretty common in the RESTEasy test suite.

This is what my test looked like:
```java
public class Main {

    public static void main(final String[] args) throws Throwable {
        //System.setProperty("java.util.logging.manager", "org.jboss.logmanager.LogManager");

        final SeBootstrap.Instance instance = SeBootstrap.start(RestActivator.class)
                .toCompletableFuture().get();

        try (Client client = ClientBuilder.newClient()) {
            final var target = client.target(instance.configuration().baseUriBuilder()).path("rest");
            try (
                    Response response = target.path("info").path("add").request().accept(MediaType.TEXT_HTML_TYPE).
                            post(Entity.form(new Form().param("key", "form-name").param("value", "form-value")))
            ) {
                System.out.println(response.getStatus());
                System.out.println(response.readEntity(String.class));
            }
        }
        //Thread.currentThread().join();
        instance.stop().toCompletableFuture().get();
    }

    /**
     * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
     */
    @ApplicationPath("/rest")
    public static class RestActivator extends Application {
        @Override
        public Set<Class<?>> getClasses() {
            return Set.of(InfoResource.class);
        }
    }

    /**
     * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
     */
    @Path("/info")
    @RequestScoped
    public static class InfoResource {

        @Inject
        private UriInfo uriInfo;

        @GET
        @Produces(MediaType.TEXT_PLAIN)
        public Response uriInfo() {
            if (uriInfo == null) {
                return Response.serverError().entity("UriInfo is null").build();
            }
            return Response.ok(uriInfo.getRequestUriBuilder().build()).build();
        }

        @POST
        @Path("/add")
        @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
        @Produces(MediaType.TEXT_HTML)
        public Response createItem(@FormParam("key") String key, @FormParam("value") String value) {
            final var html = String.format("<html>Key: %s<br>Value: %s</html>", key, value);
            return Response.ok(html).build();
        }

    }
}
```

The response was:
```
200
<html>Key: form-name<br>Value: form-value</html>
```

**Comment by nicolasduminil** [comment]
Hello,
Finally it was obvious:

        <dependency>
          <groupId>org.jboss.resteasy</groupId>
          <artifactId>resteasy-client</artifactId>
          <scope>provided</scope>
        </dependency>

I've omitted the "scope provided" and I probably fell down in a class loader ambush :-). This is also the reason that, using another JAX-RS client, which weren't present on the server classpath, worked as expected.

Many thanks for your help and support.

Kind regards,

Nicolas
